internal/parse/parser_test.go

```diff
@@ -12,3 +12,73 @@ func TestInit(t *testing.T) {
 	assert.Equal(t, nil, err)
 	assert.Equal(t, (*ParsedExpr)(nil), expr)
 }
+
+func TestPeekByte(t *testing.T) {
+	p := NewParser()
+	p.Parse("")
+	assert.Equal(t, false, p.peekByte('a'))
+	p.Parse("b")
+	assert.Equal(t, false, p.peekByte('a'))
+	p.Parse("a")
+	assert.Equal(t, true, p.peekByte('a'))
+}
+
+func TestSkipByte(t *testing.T) {
+	p := NewParser()
+	p.Parse("")
+	assert.Equal(t, false, p.skipByte('a'))
+	p.Parse("abc")
+	assert.Equal(t, false, p.skipByte('b'))
+	p.Parse("abc")
+	assert.Equal(t, true, p.skipByte('a'))
+	assert.Equal(t, true, p.peekByte('b'))
+}
+
+func TestSkipByteFind(t *testing.T) {
+	p := NewParser()
+	p.Parse("")
+	assert.Equal(t, false, p.skipByteFind('a'))
+	p.Parse("abcde")
+	assert.Equal(t, false, p.skipByteFind('x'))
+	assert.Equal(t, true, p.skipByteFind('b'))
+	assert.Equal(t, true, p.peekByte('c'))
+	p.Parse("abcde ")
+	assert.Equal(t, true, p.skipByteFind(' '))
+	assert.Equal(t, false, p.skipByteFind(' '))
+}
+
+func TestSkipSpaces(t *testing.T) {
+	p := NewParser()
+	p.Parse("")
+	assert.Equal(t, false, p.skipSpaces())
+	p.Parse("abc    d")
+	assert.Equal(t, false, p.skipSpaces())
+	p.Parse("     abcd")
+	assert.Equal(t, true, p.skipSpaces())
+	assert.Equal(t, true, p.peekByte('a'))
+	p.Parse("  \t  abcd")
+	assert.Equal(t, true, p.skipSpaces())
+	assert.Equal(t, true, p.peekByte('\t'))
+}
+
+func TestSkipString(t *testing.T) {
+	p := NewParser()
+	p.Parse("")
+	assert.Equal(t, false, p.skipString("a"))
+	p.Parse("helloworld")
+	assert.Equal(t, true, p.skipString("hElLo"))
+	assert.Equal(t, true, p.peekByte('w'))
+	p.Parse("hello world")
+	assert.Equal(t, true, p.skipString("hello"))
+	assert.Equal(t, true, p.peekByte(' '))
+}
+
+func TestIsNameByte(t *testing.T) {
+	assert.Equal(t, true, isNameByte('f'))
+	assert.Equal(t, true, isNameByte('0'))
+	assert.Equal(t, true, isNameByte('5'))
+	assert.Equal(t, false, isNameByte('à'))
+	assert.Equal(t, false, isNameByte('\n'))
+	assert.Equal(t, false, isNameByte('\r'))
+	assert.Equal(t, false, isNameByte('\t'))
```

Comment:
All of these tests feel quite repetitive. It seems worth organizing them in a table format. Something like:
```
type parseHelperTest struct {
        bytef   func(byte) bool
        stringf func(string) bool
        result  bool
        data    string
}

var parseTests = []parseHelperTest{{
        {bytef: isNameByte, result: true, data: "f"}
        {bytef: isNameByte, result: true, data: "0"}
        {bytef: isNameByte, result: true, data: "5"}
        ... etc ...
}
```
Then just have one single test function that for-loops the whole thing at once.

Reply:
I can do like in the experiment. I couple of questions though:

1. Aren't this table-like tests more difficult to debug?
```
--- FAIL: TestRunTable (0.00s)
    parser_test.go:32: 
                Error Trace:    /parser_test.go:32
                Error:          Not equal: 
                                expected: false
                                actual  : true
                Test:           TestRunTable
FAIL
```
with parser_test.go:32 being the `assert` line of the test function. There is no indication of the test that is actually failing.

2. The parser keeps a state. Do you want table-like tests for those too? I'm thinking of tests like [this](https://github.com/canonical/sqlair/pull/3/commits/7e6558f0e038167f8f7a4608a955f9d4296ec5bc#diff-acf51ab254114fd61f15d384a41705347f24cf03efacf3447fd9c42f32b35633R37). Wouldn't this over complicate the function using the test table?

Reply:
For (1), just print the test line being executed with %#v on the failure cases, so people can easily spot the failure.

For (2), fair enough, let's keep it as it is for now so we can move on. It's repetitive but it's not crazy yet, so let's wait for until/if it bothers us more.

Reply:
As of now, all the tests are in a table-like fashion.